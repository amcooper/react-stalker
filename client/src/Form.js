import React, { Component } from "react";
import './Form.css';
import config from "./config";

class Form extends Component {

  constructor(props) {
    super(props);
    let date = new Date( Date.now());
    this.state = {
      celebrity: '',
      stalker: '',
      date: `${ date.toISOString().slice( 0, -1 )}`,
      location: '',
      comment: ''
    };

    this.handleChange = this.handleChange.bind( this );
    this.handleSubmit = this.handleSubmit.bind( this );
  }

  handleChange( event ) {
    const target = event.target;
    const value = target.value;
    const name = target.name;
    this.setState({ [name]: value });
  }

  handleSubmit(event) {
    // const currentTimestamp = "2018-09-07";
    const currentTimestamp = new Date().toUTCString();
    let data = new URLSearchParams();
    data.append('celebrity', this.state.celebrity);
    data.append('stalker', this.state.stalker);
    data.append('date', this.state.date);
    data.append('location', this.state.location);
    data.append('comment', this.state.comment);
    data.append('created_at', this.props.item ? this.props.item.created_at || new Date( "1970-01-01 11:59:59" ) : currentTimestamp);
    data.append('updated_at', currentTimestamp);

    event.preventDefault();

    let fetchURL = `
      ${config.apiURL[ process.env.NODE_ENV ]}${( this.props.isEditForm ? `/${this.props.item ? this.props.item.id : ''}` : '' )}`;
    let fetchMethod = this.props.isEditForm ? 'PUT' : 'POST';

    fetch(fetchURL, {
      method: fetchMethod,
      headers: {
        'Accept': 'application/json, application/xml, text/plain, text/html, *.*',
        'Content-Type': 'application/x-www-form-urlencoded; charset=utf-8'
      },
      body: data,
      mode: 'cors'
    })
    .then((res) => {
      this.resetForm();
      this.props.resetAppState();
      this.props.getSightings();
      return res.json();
    })
    .catch((e) => {console.error(e.stack); });
  }

  resetForm() {
    let date = new Date( Date.now());
    this.setState({
      celebrity: '',
      stalker: '',
      date: `${ date.toISOString().slice( 0, -1 )}`,
      location: '',
      comment: ''
    });
  }

  componentWillReceiveProps(nextProps) {
    if ( nextProps.item ) {
      this.setState({
        celebrity: nextProps.item.celebrity || '',
        stalker: nextProps.item.stalker || '',
        date: nextProps.item.date.toISOString().slice( 0, -1 ) || '',
        location: nextProps.item.location || '',
        comment: nextProps.item.comment || '',
      });
    }
  }

  render() {
    return (
      <form id='input-form' onSubmit={this.handleSubmit}>
        <div>
          <label>Celebrity:</label>
          <input type="text" name="celebrity" value={this.state.celebrity} onChange={this.handleChange} />
        </div>
        <div>
          <label>Stalker:</label>
          <input type="text" name="stalker" value={this.state.stalker} onChange={this.handleChange} />
        </div>
        <div>
          <label>Date and time:</label>
          <input type="datetime-local" name="date" value={this.state.date} onChange={this.handleChange} />
        </div>
        <div>
          <label>Location:</label>
          <input type="text" name="location" value={this.state.location} onChange={this.handleChange} />
        </div>
        <div>
          <label>Comment:</label>
          <textarea name="comment" value={this.state.comment} onChange={this.handleChange} />
        </div>
        <input type="submit" value={this.props.isEditForm ? "Update" : "Create"} />
      </form>
    )
  }
}

export default Form;
