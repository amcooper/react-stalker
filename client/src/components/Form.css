label {
  float: left;
  text-align: right;
  margin-right: 15px;
  width: 125px;
}

#input-form {
  background-color: #07889B;
  color: white;
}

#input-form div {
  margin-bottom: 5px;
}

#input-form input, #input-form textarea {
  width: 275px;
}

#input-form input[type="submit"] {
  width: 150px;
}
