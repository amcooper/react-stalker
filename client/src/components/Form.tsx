/// <reference path="interfaces.d.ts" />

import * as React from "react";
import { Component } from "react";
import "./Form.css";

export default class Form extends Component<{}, IFormState> {
  constructor(props: {}) {
    super(props);
    let date = new Date(Date.now());
    this.state = {
      celebrity: "",
      stalker: "",
      date: `${date.toISOString().slice(0, -1)}`,
      location: "",
      comment: ""
    };

    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  handleChange(event) {
    const target = event.target;
    const value = target.value;
    const name = target.name;
    this.setState({ [name]: value });
  }

  handleSubmit(event) {
    let data = new URLSearchParams();
    data.append("celebrity", this.state.celebrity);
    data.append("stalker", this.state.stalker);
    data.append("date", this.state.date);
    data.append("location", this.state.location);
    data.append("comment", this.state.comment);

    event.preventDefault();

    let fetchURL = `/api/v1/sightings${
      this.props.isEditForm
        ? `/${this.props.item ? this.props.item.id : ""}`
        : ""
    }`;
    let fetchMethod = this.props.isEditForm ? "PUT" : "POST";

    fetch(fetchURL, {
      method: fetchMethod,
      headers: {
        Accept: "application/json, application/xml, text/plain, text/html, *.*",
        "Content-Type": "application/x-www-form-urlencoded; charset=utf-8"
      },
      body: data,
      mode: "cors"
    })
      .then(res => {
        this.resetForm();
        this.props.resetAppState();
        this.props.getSightings();
        return res.json();
      })
      .catch(e => {
        console.error(e.stack);
      });
  }

  resetForm() {
    let date = new Date(Date.now());
    this.setState({
      celebrity: "",
      stalker: "",
      date: `${date.toISOString().slice(0, -1)}`,
      location: "",
      comment: ""
    });
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.item) {
      this.setState({
        celebrity: nextProps.item.celebrity || "",
        stalker: nextProps.item.stalker || "",
        date: nextProps.item.date.toISOString().slice(0, -1) || "",
        location: nextProps.item.location || "",
        comment: nextProps.item.comment || ""
      });
    }
  }

  render() {
    return (
      <form id="input-form" onSubmit={this.handleSubmit}>
        <div>
          <label>Celebrity:</label>
          <input
            type="text"
            name="celebrity"
            value={this.state.celebrity}
            onChange={this.handleChange}
          />
        </div>
        <div>
          <label>Stalker:</label>
          <input
            type="text"
            name="stalker"
            value={this.state.stalker}
            onChange={this.handleChange}
          />
        </div>
        <div>
          <label>Date & time:</label>
          <input
            type="datetime-local"
            name="date"
            value={this.state.date}
            onChange={this.handleChange}
          />
        </div>
        <div>
          <label>Location:</label>
          <input
            type="text"
            name="location"
            value={this.state.location}
            onChange={this.handleChange}
          />
        </div>
        <div>
          <label>Comment:</label>
          <textarea
            name="comment"
            value={this.state.comment}
            onChange={this.handleChange}
          />
        </div>
        <input
          type="submit"
          value={this.props.isEditForm ? "Update" : "Create"}
        />
      </form>
    );
  }
}
