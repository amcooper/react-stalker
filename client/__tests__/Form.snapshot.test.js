import React from "react";
import Form from "../src/Form";
import { MemoryRouter } from "react-router-dom";
import renderer from "react-test-renderer";
import { stalkList } from "../fixtures/fixtures";

describe("Form component renders the form correctly", () => {
  it("create form", () => {
    Date.now = jest.fn(() => 1537963884615);
    const rendered = renderer.create(<Form isEditForm={false} item={null} />);

    expect(rendered.toJSON()).toMatchSnapshot();
  });

  it("edit form", () => {
    const rendered = renderer.create(
      <Form isEditForm={true} item={stalkList[0]} />
    );

    expect(rendered.toJSON()).toMatchSnapshot();
  });
});
