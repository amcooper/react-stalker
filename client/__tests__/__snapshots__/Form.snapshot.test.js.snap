// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form component renders the form correctly create form 1`] = `
<form
  id="input-form"
  onSubmit={[Function]}
>
  <div>
    <label>
      Celebrity:
    </label>
    <input
      name="celebrity"
      onChange={[Function]}
      type="text"
      value=""
    />
  </div>
  <div>
    <label>
      Stalker:
    </label>
    <input
      name="stalker"
      onChange={[Function]}
      type="text"
      value=""
    />
  </div>
  <div>
    <label>
      Date and time:
    </label>
    <input
      name="date"
      onChange={[Function]}
      type="datetime-local"
      value="2018-09-26T12:11:24.615"
    />
  </div>
  <div>
    <label>
      Location:
    </label>
    <input
      name="location"
      onChange={[Function]}
      type="text"
      value=""
    />
  </div>
  <div>
    <label>
      Comment:
    </label>
    <textarea
      name="comment"
      onChange={[Function]}
      value=""
    />
  </div>
  <input
    type="submit"
    value="Create"
  />
</form>
`;

exports[`Form component renders the form correctly edit form 1`] = `
<form
  id="input-form"
  onSubmit={[Function]}
>
  <div>
    <label>
      Celebrity:
    </label>
    <input
      name="celebrity"
      onChange={[Function]}
      type="text"
      value=""
    />
  </div>
  <div>
    <label>
      Stalker:
    </label>
    <input
      name="stalker"
      onChange={[Function]}
      type="text"
      value=""
    />
  </div>
  <div>
    <label>
      Date and time:
    </label>
    <input
      name="date"
      onChange={[Function]}
      type="datetime-local"
      value="2018-09-26T12:11:24.615"
    />
  </div>
  <div>
    <label>
      Location:
    </label>
    <input
      name="location"
      onChange={[Function]}
      type="text"
      value=""
    />
  </div>
  <div>
    <label>
      Comment:
    </label>
    <textarea
      name="comment"
      onChange={[Function]}
      value=""
    />
  </div>
  <input
    type="submit"
    value="Update"
  />
</form>
`;
